File: sdks/canary/typescript/scenarios/d-session-subtask.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// D-SESSION-SUBTASK canary — TypeScript SDK

import { LLMSafeSpaces } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch, waitActive, ensureSessionWithRetry, sleep } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  if (!cfg.llmApiKey) { r.ok('session-subtask: skipped (no LLM API key)'); return; }
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 180000, fetch: nodeFetch as any });
  let wsId: string | null = null;
  try {
    const [ok, ws] = await r.assertNoError(
      () => c.workspaces.create({ name: 'canary-ts-subtask', runtime: 'base', storageSize: '1Gi' }),
      'create: no error');
    if (!ok || !ws) return;
    wsId = ws.id;

    const phase = await waitActive(c, wsId);
    r.assert(phase === 'Active', 'reach-active', `got "${phase}"`);
    if (phase !== 'Active') return;

    const sess = await ensureSessionWithRetry(c, wsId, 5);
    r.ok('ensure-session: no error');
    const sid = sess.sessionId;

    const [ok2, msg] = await r.assertNoError(
      () => c.sessions.sendMessage(wsId!, sid,
        'Use the task tool to spawn a subagent that responds with exactly: SUBTASK-OK'),
      'send-subtask-message: no error');
    if (ok2 && msg) r.assert(msg.content.length > 0, 'send-subtask-message: non-empty content');

    let subtaskFound = false;
    const deadline = Date.now() + 30000;
    while (Date.now() < deadline && !subtaskFound) {
      await sleep(3000);
      const [ok3, sessions] = await r.assertNoError(
        () => c.sessions.list(wsId!), 'poll-sessions: no error');
      if (ok3 && sessions) {
        for (const s of sessions as any[]) {
          if (s.parentId && s.parentId.length > 0) {
            subtaskFound = true;
            r.assert(s.parentId === sid, 'subtask: parentId matches top-level session',
              `${s.parentId} vs ${sid}`);
            break;
          }
        }
      }
    }

    if (!subtaskFound) {
      r.ok('subtask: model did not spawn subagent (skipped gracefully)');
    }

    const [ok4, sessions] = await r.assertNoError(
      () => c.sessions.list(wsId!), 'final-session-list: no error');
    if (ok4 && sessions) {
      const topLevel = (sessions as any[]).find(s => s.id === sid);
      if (topLevel) {
        r.assert(!topLevel.parentId || topLevel.parentId === null || topLevel.parentId === '',
          'top-level: no parentId', `got "${topLevel.parentId}"`);
      }
    }

  } finally {
    if (wsId) { try { await c.workspaces.delete(wsId); } catch {} }
  }
}

async function main() {
  const r = new Runner('session-subtask');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });

File: sdks/canary/python/scenarios/d_session_subtask.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""D-SESSION-SUBTASK canary — Python SDK"""

from __future__ import annotations

import sys
import os
import time

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import (
    Runner,
    Config,
    config_from_env,
    wait_active,
    ensure_session_with_retry,
)
from llmsafespaces import LLMSafeSpaces
from llmsafespaces.client import message_text


def run(r: Runner, cfg: Config) -> None:
    if not cfg.llm_api_key:
        r.ok("session-subtask: skipped (no LLM API key)")
        return

    c = LLMSafeSpaces(cfg.api_url, api_key=cfg.api_key, timeout=120.0)
    ws_id = None
    try:
        ok, ws = r.assert_no_error(
            lambda: c.workspaces.create(
                name="canary-py-subtask", runtime="base", storage_size="1Gi"
            ),
            "create: no error",
        )
        if not ok:
            return
        ws_id = ws.id

        phase = wait_active(c, ws_id)
        r.assert_(phase == "Active", "reach-active", f"got {phase!r}")
        if phase != "Active":
            return

        try:
            sess = ensure_session_with_retry(c, ws_id, 5)
        except Exception as e:
            r.fail("ensure-session: no error", str(e))
            return
        r.ok("ensure-session: no error")
        sid = sess.sessionId

        ok2, msg = r.assert_no_error(
            lambda: c.sessions.send_message(
                ws_id,
                sid,
                "Use the task tool to create a subtask that writes hello to /tmp/canary-sub.txt",
            ),
            "send-message: no error",
        )
        if ok2 and msg is not None:
            r.assert_(len(message_text(msg)) > 0, "send-message: non-empty text")

        subtask_found = False
        deadline = time.time() + 30
        while time.time() < deadline:
            try:
                sessions = c.sessions.list(ws_id)
                for s in sessions:
                    if s.get("parentId") and s["parentId"] != "":
                        subtask_found = True
                        r.assert_(
                            len(s["parentId"]) > 0,
                            "subtask: has parentId",
                            repr(s["parentId"]),
                        )
                        break
            except Exception:
                pass
            if subtask_found:
                break
            time.sleep(2)

        if not subtask_found:
            r.ok("subtask: skipped (model does not use task tool)")

    finally:
        if ws_id:
            try:
                c.workspaces.delete(ws_id)
            except Exception:
                pass


if __name__ == "__main__":
    r = Runner("session-subtask")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())
